File: app/src/main/java/com/saeed/chandwidget/util/Formatter.java
package com.saeed.chandwidget.util;

import com.saeed.chandwidget.data.PriceItem;
import java.text.DecimalFormat;
import java.text.DecimalFormatSymbols;
import java.util.Locale;

public class Formatter {

    private static final DecimalFormatSymbols EN = new DecimalFormatSymbols(Locale.US);
    private static final DecimalFormatSymbols FA;

    static {
        FA = new DecimalFormatSymbols(new Locale("fa", "IR"));
        FA.setGroupingSeparator('،');
        FA.setDecimalSeparator('.');
    }

    public static String formatPrice(double price, PriceItem.PriceType type, boolean persian) {
        switch (type) {
            case CURRENCY_TOMAN: return formatToman(price, persian);
            case GOLD_TOMAN:     return formatGoldToman(price, persian);
            case CRYPTO_USD:
            case GOLD_USD:       return formatCryptoUsd(price, persian);
            default:             return String.valueOf((long) price);
        }
    }

    private static String formatToman(double price, boolean persian) {
        DecimalFormat df = new DecimalFormat("#,###", persian ? FA : EN);
        return df.format((long) price);
    }

    private static String formatGoldToman(double price, boolean persian) {
        double million = price / 1_000_000.0;
        if (million >= 1.0) {
            DecimalFormat df = new DecimalFormat("#,##0.00", persian ? FA : EN);
            return df.format(million) + (persian ? "م.ت" : "M");
        } else {
            DecimalFormat df = new DecimalFormat("#,##0.0", persian ? FA : EN);
            return df.format(price / 1000.0) + (persian ? "ه.ت" : "K");
        }
    }

    private static String formatCryptoUsd(double price, boolean persian) {
        if (price >= 1000) {
            DecimalFormat df = new DecimalFormat("#,##0", persian ? FA : EN);
            return "$" + df.format((long) price);
        } else if (price >= 1) {
            DecimalFormat df = new DecimalFormat("#,##0.00", persian ? FA : EN);
            return "$" + df.format(price);
        } else {
            DecimalFormat df = new DecimalFormat("0.0000", persian ? FA : EN);
            return "$" + df.format(price);
        }
    }

    public static String formatChange(double change, PriceItem.PriceType type, boolean persian) {
        if (change == 0) return "—";
        String sign = change > 0 ? "↑" : "↓";
        double abs = Math.abs(change);
        String val;
        switch (type) {
            case GOLD_TOMAN:
            case CURRENCY_TOMAN:
                if (abs >= 1_000_000)     val = fmt(abs / 1_000_000) + "M";
                else if (abs >= 1000)      val = fmt(abs / 1000) + "K";
                else                       val = fmt(abs);
                break;
            case CRYPTO_USD:
            case GOLD_USD:
                if (abs >= 1000)           val = "$" + fmt(abs / 1000) + "K";
                else                       val = "$" + fmt(abs);
                break;
            default:
                val = fmt(abs);
        }
        return sign + val;
    }

    private static String fmt(double v) {
        return new DecimalFormat("#,##0.##", EN).format(v);
    }

    /**
     * True if the string contains any Arabic/Persian-script letters.
     *
     * Used to decide per-item text alignment instead of trusting the app's
     * overall Persian/English mode: a few registry entries (e.g. EOS) have no
     * Farsi translation and keep their Latin ticker in nameFa, so checking
     * the app language alone isn't enough to know whether a given label
     * actually needs right-to-left alignment.
     */
    public static boolean containsRtl(String s) {
        if (s == null) return false;
        for (int i = 0; i < s.length(); i++) {
            char c = s.charAt(i);
            if ((c >= 0x0590 && c <= 0x08FF) || (c >= 0xFB1D && c <= 0xFDFF) || (c >= 0xFE70 && c <= 0xFEFF)) {
                return true;
            }
        }
        return false;
    }

    public static String toPersianDigits(String s) {
        StringBuilder sb = new StringBuilder();
        for (char c : s.toCharArray()) {
            if (c >= '0' && c <= '9') sb.append((char)(c - '0' + '\u06F0'));
            else sb.append(c);
        }
        return sb.toString();
    }

    /**
     * A compact label for narrow widget rows (2×2 3-row).
     *
     * In English mode this is just the 3-letter code (USD, EUR, ...), same as
     * before — that's already as compact as it gets and is a recognized
     * standard, so there's no real gain from showing the full name there.
     *
     * In Persian mode, the old code also fell back to the Latin code ("USD")
     * to avoid squeezing long Farsi names like "نیم سکه" into a fixed-width
     * column. That's a legitimate rendering constraint — the widget is only
     * ~110dp — but showing Latin abbreviations in an otherwise all-Farsi UI
     * reads as inconsistent, not "abbreviated". Gold/coin items already have
     * a short, correct Farsi word in symbolFa ("طلا", "سکه", "نیم", "ربع").
     * Currencies didn't have an equivalent short Farsi form, so this takes
     * the first word of the full Farsi name ("دلار آمریکا" → "دلار",
     * "پوند انگلیس" → "پوند"), which is how a person would actually
     * abbreviate it, rather than switching alphabets.
     *
     * Note this doesn't create full-name-fits-in-3-rows: it swaps one
     * abbreviation for a better one. Two currencies sharing a first word
     * (e.g. "دلار آمریکا" / "دلار استرالیا" both start with "دلار") will look
     * the same here; the flag emoji next to it is what disambiguates them.
     */
    public static String shortLabel(PriceItem item, boolean persian) {
        if (!persian) return item.getSymbolEn();
        String fa = item.getSymbolFa();
        String name = item.getNameFa();
        if (fa != null && !fa.equals(item.getSymbolEn())) return fa; // already a real Farsi word (gold/coins)
        if (name == null || name.isEmpty()) return fa;
        int sp = name.indexOf(' ');
        return sp > 0 ? name.substring(0, sp) : name;
    }
}

File: windows/chandwidget/formatter.py
"""
Direct port of Formatter.java from the Android app.
"""
from .registry import PriceType, PriceItem

_PERSIAN_GROUP_SEP = "،"
_PERSIAN_DIGITS = "۰۱۲۳۴۵۶۷۸۹"


def _group_thousands(int_str: str, sep: str) -> str:
    neg = int_str.startswith("-")
    if neg:
        int_str = int_str[1:]
    parts = []
    while len(int_str) > 3:
        parts.insert(0, int_str[-3:])
        int_str = int_str[:-3]
    parts.insert(0, int_str)
    out = sep.join(parts)
    return ("-" + out) if neg else out


def _fmt_int(value: float, persian: bool) -> str:
    s = f"{int(round(value)):d}"
    return _group_thousands(s, _PERSIAN_GROUP_SEP if persian else ",")


def _fmt_decimal(value: float, decimals: int, persian: bool) -> str:
    s = f"{value:,.{decimals}f}"
    int_part, _, frac_part = s.partition(".")
    int_part = int_part.replace(",", "")
    grouped = _group_thousands(int_part, _PERSIAN_GROUP_SEP if persian else ",")
    return f"{grouped}.{frac_part}" if decimals > 0 else grouped


def format_price(price: float, ptype: PriceType, persian: bool) -> str:
    if ptype == PriceType.CURRENCY_TOMAN:
        return _fmt_int(price, persian)
    if ptype == PriceType.GOLD_TOMAN:
        return _format_gold_toman(price, persian)
    if ptype in (PriceType.CRYPTO_USD, PriceType.GOLD_USD):
        return _format_crypto_usd(price, persian)
    return str(int(price))


def _format_gold_toman(price: float, persian: bool) -> str:
    million = price / 1_000_000.0
    if million >= 1.0:
        return _fmt_decimal(million, 2, persian) + ("م.ت" if persian else "M")
    else:
        return _fmt_decimal(price / 1000.0, 1, persian) + ("ه.ت" if persian else "K")


def _format_crypto_usd(price: float, persian: bool) -> str:
    if price >= 1000:
        return "$" + _fmt_int(price, persian)
    elif price >= 1:
        return "$" + _fmt_decimal(price, 2, persian)
    else:
        return "$" + _fmt_decimal(price, 4, persian)


def _fmt_compact(v: float) -> str:
    # Always Latin digits/format regardless of language, like Formatter.fmt() in Java
    if v == int(v):
        return f"{int(v):,}"
    s = f"{v:,.2f}".rstrip("0").rstrip(".")
    return s


def format_change(change: float, ptype: PriceType, persian: bool) -> str:
    if change == 0:
        return "—"
    sign = "↑" if change > 0 else "↓"
    abs_v = abs(change)
    if ptype in (PriceType.GOLD_TOMAN, PriceType.CURRENCY_TOMAN):
        if abs_v >= 1_000_000:
            val = _fmt_compact(abs_v / 1_000_000) + "M"
        elif abs_v >= 1000:
            val = _fmt_compact(abs_v / 1000) + "K"
        else:
            val = _fmt_compact(abs_v)
    elif ptype in (PriceType.CRYPTO_USD, PriceType.GOLD_USD):
        if abs_v >= 1000:
            val = "$" + _fmt_compact(abs_v / 1000) + "K"
        else:
            val = "$" + _fmt_compact(abs_v)
    else:
        val = _fmt_compact(abs_v)
    return sign + val


def contains_rtl(s: str) -> bool:
    if not s:
        return False
    for c in s:
        cp = ord(c)
        if (0x0590 <= cp <= 0x08FF) or (0xFB1D <= cp <= 0xFDFF) or (0xFE70 <= cp <= 0xFEFF):
            return True
    return False


def to_persian_digits(s: str) -> str:
    out = []
    for c in s:
        if "0" <= c <= "9":
            out.append(_PERSIAN_DIGITS[ord(c) - ord("0")])
        else:
            out.append(c)
    return "".join(out)


def short_label(item: PriceItem, persian: bool) -> str:
    """Compact label for narrow widget rows (2x2 3-row) — mirrors Formatter.shortLabel()."""
    if not persian:
        return item.symbol_en
    fa = item.symbol_fa
    name = item.name_fa
    if fa and fa != item.symbol_en:
        return fa
    if not name:
        return fa
    sp = name.find(" ")
    return name[:sp] if sp > 0 else name
